Add bindings for WebKitURISchemeRequest.

aartaka opened the pull request:
This is a second attempt at #46. Reasons:
- I have accidentally deleted the fork :sweat_smile: 
- I've understood that `g_memory_input_stream_new_from_data` is the only `GInputStream` function that we need and it can be easily `flet`-bound.

# Things To Discuss
- Maybe binding `g_memory_input_stream_new_from_data` in `defcfun`? 
  + It's safer than `cffi:foreign-funcall`.
   - It can become irrelevant if `cl-cffi-gtk` gets `GInputStream` bindings, though.
- How do we enforce the multi-valued returns in callbacks? This change has too much implicit return values conventions... 
- `with-g-error` doesn't fit `webkit-uri-scheme-request-finish-error` because we need to create `GError` and pass it to this function, instead of getting the error from elsewhere (`with-g-error` is made for that). That's quite a quirky design that I'm uncertain about.

# How Has This Been Tested:

- See `custom-scheme-browser-main` in demo/simple-browser.lisp. 
  - It doesn't work yet, I'm trying to figure it out. That's a draft of how the scheme addition should look.

How does the new approach looks?

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,95 @@
+    (flet ((g-memory-input-stream-new-from-data (data-string)
+             (let* ((ffi-string (cffi:foreign-string-alloc data-string))
+                    ;; TODO: Can we rely on lisp's `length' here?
+                    ;; Will it give the same result as strlen on ffi-string?
```
Unsure, but I think `length' understands unicode, so maybe you'd need to coerce this to a byte array first.

aartaka replied:
Solution was quite simple: `cffi:foreign-string-alloc` returns the length of the resulting string as a second value :)

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,95 @@
+            ;; TODO: Use `unwind-protect' to free the allocated objects?
+            (when (callback-function callback)
+              (multiple-value-bind (data-type data)
+                  ;; Callback function should return data type (e.g., "text/html") as a first value
```
no comma after "e.g."

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,95 @@
+                    (g-memory-input-stream-new-from-data data)
+                  (webkit-uri-scheme-request-finish request stream data-length data-type)
+                  (g:g-object-unref stream)
+                  ;; That's why g-memory-input-stream-new-from-data returns the string
```
Why?  Typo?

aartaka replied:
No, `g-memory-input-stream-new-from-data` returns the `ffi-string` for it to be freed later. Maybe it's not worth explaining twice, though.

Ambrevar replied:
Oh, the comment was about the `g-object-unref` _above_, right?
I didn't get that.  To avoid this kind of ambiguity, I suggest writing comments _above_ the target code, ending with `:` to make it completely clear.

aartaka replied:
No, it's about `cffi:foreign-string-free` below the comment :) 

What I meant was to emphasize that we need to free this string because it's created in `g-memory-input-stream-new-from-data` and cannot be freed there. This idea, however, should be understandable from a single comment inside the `g-memory-input-stream-new-from-data`. This comment we discuss is unnecessary and confusing. I'll remove it.

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,95 @@
+                                                  ;; TODO: This should ideally be g_free()
+                                                  ;; to free the ffi-string automatically.
+                                                  :pointer (cffi:null-pointer))))
+               ;; ffi-string is returned because we need to free it afterwards
```
Missing sentence period.

Ambrevar commented:
I just had a cursory look, looks good to me so far.
I'll have a closer look when I find the time :)

About your questions:
- I think the flet is fine for now.  We can factor it if we ever meet another occurrence.  I also suggest opening an issue upstream about the missing binding.  Finally, if you are going to leave the flet, please add a TODO comment explaining why you've implemented it as an flet.

- > How do we enforce the multi-valued returns in callbacks? This change has too much implicit return values conventions...
  What do you mean with "enforce"?  You mean have something more structured, with better documentation and not just comments?  What about returning a structure instead?

- `g-error`: I don't know, I'll need to have a closer look.

aartaka commented:
Got back to this and cleaned the code up.

Still no luck in making custom URLs work, but it does at least display the right error message ("The custom url request for URI hello:stranger failed"). 

What have changed:
- `webkit-uri-scheme-request` is a GObject now.
- definitions for `g-input-stream` and `g-memory-input-stream` along with an extreacted defcfun for `g_memory_input_stream_new_from_data`.
- `callback` has `web-view` instead of `object` again. It errored out with `object`.

I'm experiensing ASDF compile errors on `package.lisp`, erros about FASL-loading and foreign libraries not being found, although accepting these makes `cl-webkit2` load just fine. Does that have to do something with my system?

Testing and review is welcome, as I don't trust my own results xD

Ambrevar commented:
Thanks for the update, I'll take a closer look hopefully soon!

About your issue:

> I'm experiensing ASDF compile errors on `package.lisp`, erros about FASL-loading and foreign libraries not being found, although accepting these makes `cl-webkit2` load just fine. Does that have to do something with my system?

Did you recently run `guix gc`?
If so, you might have a grafting issue.  Try building your Common Lisp
profile with `--no-grafts`.
If you use the cl-packages, delete your .fasls and try again.

aartaka commented:
> Did you recently run `guix gc`? 

Oh God, I didn't :sweat: 

> If so, you might have a grafting issue. Try building your Common Lisp profile with `--no-grafts`. If you use the cl-packages, delete your .fasls and try again.

This I have already tried, but `guix gc` seems to be the thing I need, thanks!

Ambrevar commented:
Misunderstanding: `guix gc` is what might have _triggered_ your issue,
not what can solve it! :/

aartaka commented:
Errors stopped appearing after I ran `guix gc` and cleared the `.cache/common-lisp/` 0_o

Ambrevar commented:
I don't think `guix gc` has something to do with it`.
However, cleaning `.cache/common-lisp` certainly does.

The reason behind this is that Guix Common Lisp packages have issues
with grafting: https://issues.guix.gnu.org/issue/33848.

Long story short, paths are stored in UCS-4 in .fasls and Lisp binaries.
The Guix grafting code only finds ASCII and UTF-8 path, so it fails to
graft Common Lisp packages properly.

When packages are grafted and the ungrafted packages are missing
(e.g. because of `guix gc`) when the .fasl / Lisp binaries will have
references to non-existing paths.

aartaka commented:
I've shortened the code, fixed type mismatch errors, and added verbose error printing.

It doesn't error anymore, but it doesn't load the page either -- just a blank view. Investigating.

Ambrevar commented:
Artyom Bologov ***@***.***> writes:

> Solution was quite simple: `cffi:foreign-string-alloc` returns the length of the resulting string as a second value :)

Good catch!

jmercouris commented:
Like Artyom, I've also been unable to succesfully create a `GInputStream` that can be handled by `webkit-uri-scheme-request-finish`.

Despite not being able to correctly call `webkit-uri-scheme-request-finish`, we still have a completely servicable implementation. I may merge this and apply it in the context of handling the `nyxt:` internal scheme. Using this patch, we should be able to show pages like `nyxt:list-buffers`, `nyxt:list-messages`, relatively safely.

To expand upon this:

The API typically expects that the following should work:
```
(webkit:webkit-web-context-register-uri-scheme-callback
       context "hello"
       #'(lambda (request)
           (format nil "<html><body><p>Hello, ~:(~a~)!</p></body></html>"
                   (webkit:webkit-uri-scheme-request-get-path request))))
```

In the above, whenever "hello://" is accessed, our callback is called, and the result of our callback is used to set the HTML of our buffer.

Since our HTML is not getting properly handled in the callback, what we can do instead is:

```
(webkit:webkit-web-context-register-uri-scheme-callback
       context "hello"
       #'(lambda (request)
           (set-html-buffer (buffer-that-made-the-request) "<html>some HTML depending on the relevant URL</html>")))
```

Until the discussion is further settled with regards to `lisp://` schemes, I will not make any changes in that regard.

jmercouris commented:
I'm sorry if it was unclear, I am saying that is what /should/ happen, the HTML /should/ be set- but since it is not, we can make a new buffer with the correct HTML. I tried to examine the contents of the `GInputStream` to make sure they are valid, but didn't have much luck. It appears they are being correctly made.

I will ask @Ferada more about this, @Crategus is not easy to get ahold of.

aartaka commented:
Marked it as ready for review because most of stuff is done. We still need to fix `GInputStream`s, though.

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,105 @@
+    (:superclass g-input-stream
+     :export nil
+     :type-initializer "g_memory_input_stream_get_type")
+    ())
```
I guess this belongs to cl-cffi-gtk, right?

If so I suggest moving these two to a separate file and add a comment explaining why we had to add them, and that ideally we should upstream them in the future.

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,105 @@
+  (%webkit-uri-scheme-request-finish-error
+   request (glib::%g-error-new-literal
+            +webkit-plugin-error+
+            4
```
What's this magic number?

aartaka replied:
This is the `:webkit-plugin-error-connection-cancelled`. It's used this way because `glib::%g-error-new-literal` does not accept keywords as enum elements, so we need to use integers.

I'll add the comment explaining this.

Ambrevar replied:
Better, define a CL constant, e.g. `+webkit-plugin-error-connection-cancelled+`.

aartaka replied:
Done, see 2fa25f6 :)

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,105 @@
+                (cffi:foreign-string-free ffi-string)))
+          (error (c)
+            (webkit-uri-scheme-request-finish-error
+             request (format nil "The custom url request for URI ~S failed with ~A: ~A"
```
Nit: We usually use lower case format characters, i.e. `~s` and `~a`, no?  Not sure what's preferred.

jmercouris replied:
I believe they have different meaning based on case.

Ambrevar replied:
I think not, the CLHS page is uncased: http://www.lispworks.com/documentation/HyperSpec/Body/22_cda.htm.

aartaka replied:
Downcased!

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,105 @@
+            (when (and callback (callback-error-function callback))
+              (funcall (callback-error-function callback) c))))))))
+
+(defun webkit-web-context-register-uri-scheme-callback (context scheme &optional call-back error-call-back)
```
Maybe this one deserves a docstring considering it's not part of webkitgtk API?

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -70,8 +70,6 @@
                                               (user-data :pointer))
   (let ((callback (find (cffi:pointer-address user-data) callbacks :key (function callback-id))))
-    (setf callbacks (delete callback callbacks))
-    ;; TODO: Use `unwind-protect' to free the allocated objects?
```
Is this TODO addressed?

aartaka replied:
Now it is, with cbcb2f7 :)

aartaka commented:
Another reason why `GMemoryInputStream` creation may fail: the definition of `g-memory-input-stream` lacks the reference to the interfaces implemented (`GSeekable` and `GPollableInputStream`). 

I remember `g-memory-stream` definition erroring out when I tried to add names of interfaces it implements. I'll try again, this time maybe defining the missing interfaces (some are not defined in `cl-cffi-gtk`).

aartaka commented:
Interfaces are there. Still blank pages for the custom scheme :(

aartaka commented:
Tried testing our `g-memory-input-stream` with `g_input_stream_read`. It works as intended. Seems like definitions are right.

aartaka commented:
You just look at that! It works with `"text/plain"`! I've changed the testing code and it shows the appropriate text :D 

So custom schemes do work. It's just that they don't show HTML somewhy...

Ambrevar commented:
My guess is that WebKit has some logic to know what to do depending on
the scheme.

For instance, when loading a file:// scheme is does not necessarily try
to render the HTML.  Maybe there is some parameter to set.

aartaka commented:
It may also be about the size of the foreign string and/or streams that we use. I had question mark glyphs displayed after the message when testing the plaintext callback.

Ambrevar commented:
Aha!

Looks like some buffer size issue!

That or the encoding: remember that CL strings may be encoded using wide
characters (it's implementation dependent).

So we if we the length from the CL string, we may end up with a longer
buffer than needed.

We might have to leverage CFFI strings or some CL library here.

jmercouris commented:
I could have sworn I tried with CFFI to make the strings...! Well, I also don't understand why it wouldn't just display the HTML as plaintext. Thanks for the update Artyom.

aartaka commented:
> Aha! Looks like some buffer size issue! 

Yes, that could well be the reason.

> That or the encoding: remember that CL strings may be encoded using wide characters (it's implementation dependent). So we if we the length from the CL string, we may end up with a longer buffer than needed. We might have to leverage CFFI strings or some CL library here.

We already do, don't we? `cffi:foreign-string-alloc` is used in `uri-scheme-processed`. It's just that the GInputStream API/WebKit can have it's own idea of length of the string.

aartaka commented:
It was a terrible typo -- `"text-html"` instead of `"text/html"`, and it was killing an otherwise working implementation :D

aartaka commented:
And, of course, there was an off-by-one error in stream size calculation. Should be resolved now that we delegate size calculation to `g-input-stream-new-from-data` (`-1` in arguments).

aartaka commented:
All working! ＼（＾▽＾）／

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,103 @@
+                (cffi:foreign-string-free ffi-string)))
+          (error (c)
+            (webkit-uri-scheme-request-finish-error
+             request (format nil "The custom url request for URI ~a failed with ~a: ~a"
```
I don't understand this sentence. Shouldn't it be

"The custom request for URI ~a failed"

?

aartaka replied:
It could but I needet it to be more verbose about errors, listing both it's type and error message. Still looks alright to me — it gives more details.

Ambrevar replied:
Misunderstanding: I didn't mean to remove that last `~a`, only to remove "url" after "custom".

Ambrevar commented on webkit2/webkit2.uri-scheme-request.lisp:
```diff
@@ -0,0 +1,103 @@
+
+(defun webkit-web-context-register-uri-scheme-callback (context scheme &optional call-back error-call-back)
+  "Register the custom scheme.
+Hide all the unpretty details (callbacks, WebKit functions, C objects
```
unpretty -> dirty

Ambrevar commented:
Tested and approved!  Shall we merged then?

jmercouris commented:
too soon before 2.0.0! As long as we don't update the submodule commit, it is fine

aartaka commented:
Yes, let's merge! Do you mind fixing the review points yourself? — I'm far from my machine :(

aartaka commented:
@jmercouris, but it's essential for 2.0 :)

jmercouris commented:
I'll handle it. No problem!

jmercouris commented:
It is not essential!!! Ay yai yai, I wish I had nx-kaomoji to express myself properly :-D

aartaka commented:
We have agreed that we must fix the `lisp:` scheme before 2.0, because it's a huge potential vulnerability. Haven't we?

Ambrevar commented:
I agree with Artyom.  The change is very simple on Nyxt end: register
the lisp:// scheme and remove the evaluating code from the resource
query handler.

jmercouris commented:
If you must! I wash my hands!

jmercouris commented:
Perhaps I am being /too/ cautious!

aartaka commented:
> Perhaps I am being /too/ cautious!

Not really :)
